File: C++/Stacks & Queues/Balanced_Brackets.cpp
// Balanced Brackets using concept of Stack
// Author: Pavan Kumar Paluri
// Time and Space Complexity; O(N) and O(N)
// Algo Expert Question: https://www.algoexpert.io/questions/Balanced%20Brackets

using namespace std;

// O(N) time and O(N) space 
bool balancedBrackets(string str) {
  // Write your code here.
	unordered_map<char, char>h_map{ {']','['},
																				{')','('},
																				{'}','{'} };
	unordered_set<char>hash_set = {'[',']','{','}','(',')'};
	vector<char>stack (1, '*'); // init stack with a single '*' char
	for(int i=0; i<str.size(); i++)
	{
		// if element not present in hash_set 
		if((hash_set.find(str[i])!=hash_set.end()) && (h_map.find(str[i])==h_map.end()))
			 {
				 stack.push_back(str[i]);
			 }
		else if((hash_set.find(str[i])!=hash_set.end()) && (h_map.find(str[i])!=h_map.end()))
			 {
				 if(stack.back() == h_map[str[i]])
				 {
					 stack.pop_back();
				 }
				 else
					 return false;
			 }
	}
  return stack.back() == '*';
}

File: Python/Stacks and Queues/Balanced_Brackets.py
# Balanced Brackets using Stacks
# Time and Space Complexity: O(N) and O(N)
# Author: Pavan Kumar Paluri
# Algo Expert Question: https://www.algoexpert.io/questions/Balanced%20Brackets

def balancedBrackets(string):
    # Write your code here.
	h_map = {']':'[', '}':'{', ')':'('}
	accept_chars = {'{','}','[',']','(',')'}
	stack = ['*']
	for i in range(0, len(string)):
		if string[i] in accept_chars and string[i] not in h_map:
			# simply insert
			stack.append(string[i])
		elif string[i] in accept_chars and string[i] in h_map: # if in h_map
			if stack[-1] == h_map[string[i]]:
				# well and good
				stack.pop()
			else:
				return False
	return stack[-1] == '*'
